lib/devise-security/models/secure_validatable.rb

```diff
@@ -76,14 +76,11 @@ def current_equal_password_validation
       end
 
       def email_not_equal_password_validation
+        return if self.class.allow_passwords_equal_to_email
         return if password.blank? || email.blank? || (!new_record? && !will_save_change_to_encrypted_password?)
-        dummy = self.class.new.tap do |user|
-          user.password_salt = password_salt if respond_to?(:password_salt)
-          # whether case_insensitive_keys or strip_whitespace_keys include email or not, any
-          # variation of the email should not be a supported password
-          user.password = email.downcase.strip
+        if ActiveSupport::SecurityUtils.secure_compare(password.downcase.strip, email.downcase.strip)
```

Comment:
@olbrich note that I ended up using `secure_compare` and not `fixed_length_secure_compare` because `fixed_length_secure_compare` raises an exception if the lengths aren't the same whereas `secure_compare` returns `false` (which is what we want).

Reply:
@dillonwelch is it possible to set the password to the email first and then update the record with the email and thus successfully bypass this check?

Reply:
To make sure I understand, you're thinking something like this (skipping confirmation for brevity)?
```ruby
user = User.create(email: 'bob@aol.com', password: 'whatever')
user.update(password: 'jim@aol.com')
user.update(email: 'jim@aol.com')
```

Reply:
Similar.  Maybe something like.
```
user = User.create(email: 'bob@aol.com', password: 'foo@example.com')
user.update(email: 'foo@example.com')
```

Reply:
So I was able to test that we do not currently error out in this scenario. We don't currently run this validation if the user is not updating their password. This is because we wouldn't have the password in memory and so wouldn't be able to compare it to the email. I'm not sure of a way to get around this?

Reply:
```ruby
user = User.create!(email: 'bob@aol.com', password: 'foO1@example.com', password_confirmation: 'foO1@example.com')
user = nil
user = User.first
user.email = 'foO1@example.com'

user.valid? # true
```

Reply:
We would have to encrypt the password the same way we do when we save it and compare the hashes, which might have been what it was originally doing.

Reply:
Maybe I'm missing something, but when I revert back to the original code I'm getting different hashes. 
```ruby
(byebug) email
"foo1@example.com"
(byebug) email_was
"bob@aol.com"
(byebug) encrypted_password
"$2a$12$PapCw7vqW6Tcr7t2ADrv3.yfDJm1Y8HevAexrJRZ4C.1L6y8rsjfu"
(byebug) dummy.encrypted_password
"$2a$12$PJRF1lYAXS9tLy/RCfCp/.TAO.EMLXzPUpVL/1C0XXKt3LA.mgtyW"
(byebug) dummy.password
"foo1@example.com"
(byebug) encrypted_password_was
"$2a$12$PapCw7vqW6Tcr7t2ADrv3.yfDJm1Y8HevAexrJRZ4C.1L6y8rsjfu"
(byebug) dummy
#<TestSecureValidatable::User id: nil, username: nil, facebook_token: nil, unique_session_id: nil, email: "", password_changed_at: nil, created_at: nil, updated_at: nil, paranoid_verification_code: nil, paranoid_verified_at: nil, paranoid_verification_attempt: 0, security_question_id: nil, security_question_answer: nil, expired_at: nil, last_activity_at: nil>